beacon_node/beacon_chain/src/validator_monitor.rs

```diff
@@ -491,6 +532,48 @@ impl<T: EthSpec> ValidatorMonitor<T> {
         }
     }
 
+    /// Process the unaggregated attestations generated by the service `attester_per_slot_service`
+    /// and check if the attestation qualifies for a reward matching the flags source/atrget/head
+    fn process_unaggregated_attestations(&mut self, state: &BeaconState<T>, spec: &ChainSpec) {
+        // Define range variables
+        let current_slot = state.slot();
+        let slot = current_slot - Slot::new(UNAGGREGATED_ATTESTATION_LAG_SLOTS as u64);
+
+        // If there is an unaggregated attestation for the slot
+        if let Some(unaggregated_attestation) = self.get_unaggregated_attestation(slot) {
```

Comment:
Perhaps in this function we could do something like:

```rust
let attested_slots = self.unaggregated_attestations
  .filter(|(attestation_slot, _)| attestation_slot <= slot)
  .copied()
  .collect();
for attested_slot in attested_slots {
   let attestation = self.unaggregated_attestations.remove(attested_slot); // Need to handle `Option` without panicking.
   if attested_slot.epoch() < state.previous_epoch() {
      continue;  // Don't process this attestation, it's too old to be processed by this state.
   }
   get_attestation_participation_flag_indices(state, attestation, ...)
}
```

This would achieve two things:

- Since `process_valid_state` only gets called for *states with blocks*, it ensures that we process attestation when there have been skipped slots between blocks. (`attestation_slot <= slot` achieves this).
- Removes attestations from the map once they've been analysed so that we do one set of logs/metrics per attestation.